text/3416-feature-metadata.md

```diff
@@ -0,0 +1,159 @@
+- Feature Name: feature-metadata
+- Start Date: 2023-04-14
+- RFC PR: [rust-lang/rfcs#3416](https://github.com/rust-lang/rfcs/pull/3416)
+- Rust Issue:
+  [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+This RFC is just meta tracking information for the three following RFCs:
+
+- [Cargo feature descriptions](https://github.com/rust-lang/rfcs/pull/3485)
+- [Cargo feature deprecation](https://github.com/rust-lang/rfcs/pull/3486)
+- [Cargo feature visibility](https://github.com/rust-lang/rfcs/pull/3487)
+
+# Motivation
+
+[motivation]: #motivation
+
+Features are widely used as a way to do things like reduce dependency count,
+gate `std` or `alloc`-dependent parts of code, or hide unstable API. Use is so
+common that many larger crates wind up with tens of feature gates, such as
+[`tokio`] with 24. Despite being a first class component of crate structure,
+there are some limitations that don't have elegant solutions:
+
+- Documentation is difficult, often requiring library authors to manually manage
+  a table of descriptions
+- There is no way to deprecate old features, as a way to help crates maintain
+  semvar compliance
+- Features cannot be hidden from use in any way
+
+This RFC proposes a plan that add that information to `Cargo.toml`, solving
+these problems.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Usage is simple: features will be able to be specified as a table, instead of
+just a dependency array. This sample section of `Cargo.toml` shows new
+possibilities:
+
+```toml
+[features]
+# Current configuration will continue to work
+foo = []
+# New configurations
+bar = { enables = ["foo"], doc = "simple docstring here"}
+baz = { enables = ["foo"], public = false}
+qux = { enables = [], deprecated = true }
+quux = { enables = [], deprecated = { since = "1.2.3", note = "don't use this!" } }
+
+# Features can also be full tables if descriptions are longer
+[features.corge]
+enables = ["bar", "baz"]
+doc = """
+# corge
+
+This could be a longer description of this feature
+"""
+```
+
+The `enables` key is synonymous with the existing array, describing what other
```

Comment:
There are several terms for this.

In general, we call it "feature activation".

Targets declare that they "require" a feature.

I also want to keep our language in mind in case we move forward with [disabling features](https://github.com/rust-lang/cargo/issues/3126).  I thought it an interesting choice that bazel refers to this as [implies](https://docs.bazel.build/versions/0.26.0/cc-toolchain-config-reference.html#feature-relationships).  Our use of "requires" aligns with theirs.

Reply:
That is interesting too. Maybe a zulip poll wouldn't be a bad idea to get a rough idea of what the community would expect, I can create one

Reply:
https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Feature.20Metadata.20RFC.20Polls.20.28please.20vote!.29